In [1]:
import pandas as pd

In [2]:
advertising = pd.read_csv("advertising.csv")

In [3]:
advertising.head()

TV  Mobile  Radio  Newspaper  Sales
0  230.1    37.8   37.8       69.2   22.1
1   44.5    39.3   39.3       45.1   10.4
2   17.2    45.9   45.9       69.3    9.3
3  151.5    41.3   41.3       58.5   18.5
4  180.8    10.8   10.8       58.4   12.9

In [4]:
advertising.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TV         200 non-null    float64
 1   Mobile     200 non-null    float64
 2   Radio      200 non-null    float64
 3   Newspaper  200 non-null    float64
 4   Sales      200 non-null    float64
dtypes: float64(5)
memory usage: 7.9 KB


In [5]:
advertising.describe()

TV      Mobile       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   23.264000   30.554000   14.022500
std     85.854236   14.846809   14.846809   21.778621    5.217457
min      0.700000    0.000000    0.000000    0.300000    1.600000
25%     74.375000    9.975000    9.975000   12.750000   10.375000
50%    149.750000   22.900000   22.900000   25.750000   12.900000
75%    218.825000   36.525000   36.525000   45.100000   17.400000
max    296.400000   49.600000   49.600000  114.000000   27.000000

In [6]:
import matplotlib.pyplot as plt

In [7]:
#advertising.hist(bins=200, figsize=(20, 15))

# Train-Test Data Splitting function
## Aa function banayu che data split karva mate

In [8]:
import numpy as np
def data_split(data, test_ratio):
    np.random.seed(42)
    shuffled = np.random.permutation(len(data))
    test_set_size = int(len(data) * test_ratio)
    test_indices = shuffled[:test_set_size]
    train_indices = shuffled[test_set_size:]
    return data.iloc[train_indices], data.iloc[test_indices]

## Function call kariyu, and data split thayo ee print karayo

In [9]:
train_set, test_set = data_split(advertising, 0.2)

print(f"Train Set Length: {len(train_set)}")
print(f"Test Set Length: {len(test_set)}")
print("\nTrain Set che aa")
print(train_set)
print("\n\nTest Set che aa")
print(test_set)

Train Set Length: 160
Test Set Length: 40

Train Set che aa
        TV  Mobile  Radio  Newspaper  Sales
79   116.0     7.7    7.7       23.1   11.0
197  177.0     9.3    9.3        6.4   12.8
38    43.1    26.7   26.7       35.1   10.1
24    62.3    12.6   12.6       18.3    9.7
122  224.0     2.4    2.4       15.6   11.6
..     ...     ...    ...        ...    ...
106   25.0    11.0   11.0       29.7    7.2
14   204.1    32.9   32.9       46.0   19.0
92   217.7    33.5   33.5       59.0   19.4
179  165.6    10.0   10.0       17.6   12.6
102  280.2    10.1   10.1       21.4   14.8

[160 rows x 5 columns]


Test Set che aa
        TV  Mobile  Radio  Newspaper  Sales
95   163.3    31.6   31.6       52.9   16.9
15   195.4    47.7   47.7       52.9   22.4
30   292.9    28.3   28.3       43.2   21.4
158   11.7    36.9   36.9       45.2    7.3
128  220.3    49.0   49.0        3.2   24.7
115   75.1    35.0   35.0       52.7   12.6
69   216.8    43.9   43.9       27.2   22.3
170   50.0    11.6

# Splitting function is already available in sklearn lib😒 

In [10]:
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(advertising, test_size=0.2, random_state=42)

print(f"Train Set Length: {len(train_set)}")
print(f"Test Set Length: {len(test_set)}")
print("\nTrain Set che aa")
print(train_set)
print("\n\nTest Set che aa")
print(test_set)

Train Set Length: 160
Test Set Length: 40

Train Set che aa
        TV  Mobile  Radio  Newspaper  Sales
79   116.0     7.7    7.7       23.1   11.0
197  177.0     9.3    9.3        6.4   12.8
38    43.1    26.7   26.7       35.1   10.1
24    62.3    12.6   12.6       18.3    9.7
122  224.0     2.4    2.4       15.6   11.6
..     ...     ...    ...        ...    ...
106   25.0    11.0   11.0       29.7    7.2
14   204.1    32.9   32.9       46.0   19.0
92   217.7    33.5   33.5       59.0   19.4
179  165.6    10.0   10.0       17.6   12.6
102  280.2    10.1   10.1       21.4   14.8

[160 rows x 5 columns]


Test Set che aa
        TV  Mobile  Radio  Newspaper  Sales
95   163.3    31.6   31.6       52.9   16.9
15   195.4    47.7   47.7       52.9   22.4
30   292.9    28.3   28.3       43.2   21.4
158   11.7    36.9   36.9       45.2    7.3
128  220.3    49.0   49.0        3.2   24.7
115   75.1    35.0   35.0       52.7   12.6
69   216.8    43.9   43.9       27.2   22.3
170   50.0    11.6

# Strong Co-relation hoy to eva lable/variable ne perfect test and train data ma split karva mate


In [11]:
# from sklearn.model_selection import StratifiedShuffleSplit
# split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
# for train_index, test_index in split.split(advertising, advertising['TV']):
#     start_train_set = advertising.loc[train_index]
#     start_test_set = advertising.loc[test_index]


# Co-relation check karva mate input and output vache

In [12]:
corr_metrix = advertising.corr()
corr_metrix['Sales'].sort_values(ascending=False)

Sales        1.000000
TV           0.782224
Mobile       0.576223
Radio        0.576223
Newspaper    0.228299
Name: Sales, dtype: float64

# Scikit-learn Design



Mainly three types of objects

1. Estimators - estimate karva mate from data e.g. SimpleImputer
2 methods che
Fit - Fit the dataset and internal paremeters
Transform method
2. Transformers - this method take input and return output based on learnings from fit()
It also has convenience function called fit_transform() which fits the data and then tranform.
3. Predictors - LinerRegration Modal is and example of pridictor.
fit() and predict() are common function. It also gives score() function which will evaluate the pridictions.




# Feature Scaling

Mainly, 2 type na Scaling methods hoy:
1. Min-max scaling (Normalization)
  (value - min)/(max - min) badhi value par aa karva nu same scale ma lava mate
  Sklearn ma class che ani mate MinMaxScaler
2. Standardization
  (value - mean)/std
  Sklearn ma che class StandardScaler

In [13]:
advertising_data = train_set.drop("Sales", axis=1)
advertising_lables = train_set["Sales"].copy()

In [14]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy="median")
imputer.fit(advertising_data) 

SimpleImputer(strategy='median')

In [15]:
imputer.statistics_

array([150.65,  21.2 ,  21.2 ,  25.6 ])

In [16]:
X = imputer.transform(advertising_data)
advertising_tr = pd.DataFrame(X, columns=advertising_data.columns)
advertising_tr.describe()

TV      Mobile       Radio   Newspaper
count  160.000000  160.000000  160.000000  160.000000
mean   150.019375   22.875625   22.875625   29.945625
std     84.418857   14.805216   14.805216   20.336449
min      0.700000    0.000000    0.000000    0.300000
25%     77.750000    9.825000    9.825000   12.875000
50%    150.650000   21.200000   21.200000   25.600000
75%    218.825000   36.425000   36.425000   44.500000
max    296.400000   49.600000   49.600000  100.900000

# Pipline

In [17]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

my_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy="median")),
    ('std_scaler', StandardScaler()),
])


# upar je badhu kariyu ee tya nai karva nu ee badhu pipeline ma karva nu avde sikhva mate chale

In [18]:
advertising_num = my_pipeline.fit_transform(advertising_tr)

In [19]:
advertising_num.shape

(160, 4)

# Select a Modal 

In [20]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
# model = LinearRegression()
# model = DecisionTreeRegressor()
model = RandomForestRegressor()
model.fit(advertising_num,advertising_lables)

RandomForestRegressor()

In [21]:
some_data = advertising_data.iloc[:1]
some_lables = advertising_lables.iloc[:1]
praper_data = my_pipeline.transform(some_data)

In [22]:
model.predict(praper_data)

array([11.119])

In [23]:
list(some_lables)

[11.0]

# Evaluating the Model

In [24]:
from sklearn.metrics import mean_squared_error
advertising_predctions = model.predict(advertising_num)
lin_mse = mean_squared_error(advertising_lables, advertising_predctions)
lin_rmse = np.sqrt(lin_mse)

In [25]:
lin_mse

0.09101948749999947

In [26]:
lin_rmse

0.30169436106762

In [27]:
import pickle
pickle.dump(model, open('advertising.pkl', 'wb'))

In [28]:
from joblib import dump,load
dump(model, 'advertising.joblib')

['advertising.joblib']

# Using better evalation - Cross Validation

In [29]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(model, advertising_num, advertising_lables, scoring="neg_mean_squared_error", cv=10)
rmse_scores = np.sqrt(-scores)

In [30]:
rmse_scores

array([0.68898059, 0.79719681, 0.89275784, 0.85736526, 0.71391859,
       0.7890166 , 0.77745197, 0.83505292, 1.35146303, 0.54696772])

In [31]:
def print_score(rmse_scores):
    print("Scores: ", scores)
    print("Mean: ", scores.mean())
    print("Standard Deviation: ", scores.std())
    

In [32]:
print_score(rmse_scores)

Scores:  [-0.47469425 -0.63552275 -0.79701656 -0.73507519 -0.50967975 -0.62254719
 -0.60443156 -0.69731337 -1.82645231 -0.29917369]
Mean:  -0.7201906624999987
Standard Deviation:  0.39273447460841493


# Model Preformance 

LinearRegression
Mean:  -2.9586365170146443
Standard Deviation:  1.8383801686736772

DecisionTreeRegressor
Mean:  -1.619999999999999
Standard Deviation:  0.8454798471134597

RandomForestRegressor
Mean:  -0.7201906624999987
Standard Deviation:  0.39273447460841493

#Best model acording to very less Error is RandomForestRegressor

# Best model acording to very less Error is RandomForestRegressor

# So we will use this model for this Example

# Now build model with the help of joblib

# Testing the model on Test Data

In [38]:
advertising_test  = test_set.drop("Sales", axis=1)
advertising_test_lables = test_set["Sales"].copy()
advertising_prepared = my_pipeline.transform(advertising_test)
advertising_pridictions = model.predict(advertising_prepared)
final_mse = mean_squared_error(advertising_test_lables, advertising_pridictions)
final_rsme = np.sqrt(final_mse)

In [34]:
final_rsme

In [40]:
print(advertising_pridictions)

[17.868 21.966 20.406  6.684 23.211 13.117 22.744  9.684 11.846 15.656
  7.042  9.065 12.117  4.499 10.429 12.723  6.709 16.526 11.305 19.54
 20.045 12.687 10.344 22.327 10.054  8.977 22.377 12.828  9.988  4.798
 11.597 10.932 22.198  8.428 15.227 20.451 12.322 20.306 12.265  7.7  ]


In [36]:
print(list(advertising_test_lables))

In [41]:
advertising_test

TV  Mobile  Radio  Newspaper
95   163.3    31.6   31.6       52.9
15   195.4    47.7   47.7       52.9
30   292.9    28.3   28.3       43.2
158   11.7    36.9   36.9       45.2
128  220.3    49.0   49.0        3.2
115   75.1    35.0   35.0       52.7
69   216.8    43.9   43.9       27.2
170   50.0    11.6   11.6       18.4
174  222.4     3.4    3.4       13.1
45   175.1    22.5   22.5       31.5
66    31.5    24.6   24.6        2.2
182   56.2     5.7    5.7       29.7
165  234.5     3.4    3.4       84.8
78     5.4    29.9   29.9        9.4
186  139.5     2.1    2.1       26.6
177  170.2     7.8    7.8       35.2
56     7.3    28.1   28.1       41.4
152  197.6    23.3   23.3       14.2
82    75.3    20.3   20.3       32.5
68   237.4    27.5   27.5       11.0
124  229.5    32.3   32.3       74.2
16    67.8    36.6   36.6      114.0
148   38.0    40.3   40.3       11.9
93   250.9    36.5   36.5       72.3
65    69.0     9.3    9.3        0.9
60    53.5     2.0    2.0       21.4
84   213.5    43.0   43.0       33.8
67   139.3    14.5   14.5       10.2
125   87.2    11.8   11.8       25.9
132    8.4    27.2   27.2        2.1
9    199.8     2.6    2.6       21.2
18    69.2    20.5   20.5       18.3
55   198.9    49.4   49.4       60.0
75    16.9    43.7   43.7       89.4
150  280.7    13.9   13.9       37.0
104  238.2    34.3   34.3        5.3
135   48.3    47.0   47.0        8.5
137  273.7    28.9   28.9       59.7
164  117.2    14.7   14.7        5.4
76    27.5     1.6    1.6       20.7

In [39]:
advertising_prepared

array([[ 0.15781217,  0.59112727,  0.59112727,  1.13227461],
       [ 0.53925283,  1.68199613,  1.68199613,  1.13227461],
       [ 1.69783431,  0.36753303,  0.36753303,  0.65380096],
       [-1.64363349,  0.95023317,  0.95023317,  0.75245532],
       [ 0.83513672,  1.77007871,  1.77007871, -1.31928629],
       [-0.89025846,  0.82149709,  0.82149709,  1.12240918],
       [ 0.79354661,  1.42452397,  1.42452397, -0.13543394],
       [-1.18851892, -0.76398933, -0.76398933, -0.56951314],
       [ 0.86009078, -1.31958713, -1.31958713, -0.8309472 ],
       [ 0.29803023, -0.02545078, -0.02545078,  0.07667294],
       [-1.40835233,  0.11683646,  0.11683646, -1.36861347],
       [-1.11484502, -1.16374872, -1.16374872, -0.01211599],
       [ 1.00387371, -1.31958713, -1.31958713,  2.70581169],
       [-1.71849568,  0.47594236,  0.47594236, -1.01345777],
       [-0.12500054, -1.40766971, -1.40766971, -0.16503025],
       [ 0.23980408, -1.02146148, -1.02146148,  0.25918351],
       [-1.69591819,  0.

In [42]:
print(advertising_pridictions)

[17.868 21.966 20.406  6.684 23.211 13.117 22.744  9.684 11.846 15.656
  7.042  9.065 12.117  4.499 10.429 12.723  6.709 16.526 11.305 19.54
 20.045 12.687 10.344 22.327 10.054  8.977 22.377 12.828  9.988  4.798
 11.597 10.932 22.198  8.428 15.227 20.451 12.322 20.306 12.265  7.7  ]
